apache2/re.c

```diff
@@ -1970,6 +1970,48 @@ msre_ruleset *msre_ruleset_create(msre_engine *engine, apr_pool_t *mp) {
 
     return ruleset;
 }
+/**
+
+
+ * Creates a ruleset by copy the parent ruleset.
+ *
+ * @param mp apr pool structure
+ * @param parent_ruleset Parent's msre_ruleset
+ *
+ * return the address of created ruleset if OK.
+ * return NULL if Something went wrong.
+ */
+
+msre_ruleset *msre_ruleset_create_and_copy(msre_engine *engine, apr_pool_t *mp, msre_ruleset *parent_ruleset) {
+    msre_ruleset *child_ruleset;
+
+    child_ruleset = apr_pcalloc(mp, sizeof(msre_ruleset));
+    if (child_ruleset == NULL) return NULL;
+    child_ruleset->mp = mp;
+    child_ruleset->engine = engine;
+
+    if (parent_ruleset == NULL) {
+        child_ruleset->phase_request_headers  = apr_array_make(mp, 25, sizeof(const msre_rule *));
+        child_ruleset->phase_request_body     = apr_array_make(mp, 25, sizeof(const msre_rule *));
+        child_ruleset->phase_response_headers = apr_array_make(mp, 25, sizeof(const msre_rule *));
+        child_ruleset->phase_response_body    = apr_array_make(mp, 25, sizeof(const msre_rule *));
+        child_ruleset->phase_logging          = apr_array_make(mp, 25, sizeof(const msre_rule *));
+        return child_ruleset;
+    }
+
+    child_ruleset->phase_request_headers  = apr_array_copy(mp, parent_ruleset->phase_request_headers);
```

Comment:
are we 110% sure that apr_array_copy is equivalent to what was happening in copy_rules_phase? I find it difficult to overview whether this is true. Since it's difficult to convince ourselves of this, and the gain will be 2% (or 0% for most customers, as they have SecRuleRemoveById), is this change really worth the risk?

Reply:
When exceptions_arr->nelts <= 0, copy_rules_phase() becomes very simple code. 

When exceptions_arr->nelts <= 0, copy is always 1 and mode only can be 0 or 2. Then, copy_rules_phase() becomes

static void copy_rules_phase(...)
{
    rule_exception **exceptions;
    msre_rule **rules;
    int i, j;
    int mode = 0;

    rules = (msre_rule **)parent_phase_arr->elts;
    for(i = 0; i < parent_phase_arr->nelts; i++) {
        msre_rule *rule = (msre_rule *)rules[i];
        if (mode == 0) {
                /* Copy the rule. */
                *(msre_rule **)apr_array_push(child_phase_arr) = rule;
                if (rule->actionset->is_chained) mode = 2;
        } else {
            /* Copy the rule (it belongs to the chain we want to include. */
            *(msre_rule **)apr_array_push(child_phase_arr) = rule;
            if ((rule->actionset == NULL)||(rule->actionset->is_chained == 0)) mode = 0;
        }
    }
}

Because the function do the same things when mode == 0 and mode == 2, then copy_rules_phase() becomes

static void copy_rules_phase(...)
{
    rule_exception **exceptions;
    msre_rule **rules;
    int i, j;
    int mode = 0;

    rules = (msre_rule **)parent_phase_arr->elts;
    for(i = 0; i < parent_phase_arr->nelts; i++) {
        msre_rule *rule = (msre_rule *)rules[i];
        /* Copy the rule. */
        *(msre_rule **)apr_array_push(child_phase_arr) = rule;
    }
}

I think it is very easy to confirm that apr_array_copy() is equivalent to copy_rules_phase()

Reply:
It's a valuable comment because the gain of this optimization is only 2%, but It's very strange that SecRuleRemoveById will remove rules at the runtime, not at the compile time. Anyway, I agree with that this one is not the major bottleneck in current ModSecurity.